src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Unix.cs

```diff
@@ -33,9 +33,30 @@ internal static unsafe int ReadAtOffset(SafeFileHandle handle, Span<byte> buffer
                 // The Windows implementation uses ReadFile, which ignores the offset if the handle
                 // isn't seekable.  We do the same manually with PRead vs Read, in order to enable
                 // the function to be used by FileStream for all the same situations.
-                int result = handle.CanSeek ?
-                    Interop.Sys.PRead(handle, bufPtr, buffer.Length, fileOffset) :
-                    Interop.Sys.Read(handle, bufPtr, buffer.Length);
+                int result;
+                if (handle.SupportsRandomAccess)
+                {
+                    // Try pread for seekable files.
+                    result = Interop.Sys.PRead(handle, bufPtr, buffer.Length, fileOffset);
+                    if (result == -1)
+                    {
+                        // Some devices (such as /dev/tty) are reported as seekable by macOS but pread is unable to handle them and returns ENXIO.
+                        // Historically we were able to handle /dev/tty using read so we need to fallback to read for that case.
+                        Interop.ErrorInfo errorInfo = Interop.Sys.GetLastErrorInfo();
+                        AssertPReadPWriteKnownErrors(errorInfo.Error);
+
+                        if (errorInfo.Error == Interop.Error.ENXIO)
```

Comment:
This needs `ESPIPE` also.

Reply:
Can you describe in which case we can get ESPIPE?

Reply:
When the path you pass in is a pipe/socket.

Reply:
for example, EPIPE when using pwrite on socket:
```
socketpair(AF_UNIX, SOCK_STREAM, 0, [3, 4]) = 0
pwrite64(3, "hello", 5, 0)              = -1 ESPIPE (Illegal seek)
```

Reply:
CanSeek reports `false` for a FileStream opening a socket so in theory we will never call pwrite for that case (unless there's another distro that unexpectedly reports true). 

I'm trying to add a unit test that covers this.

Reply:
As part of https://github.com/dotnet/runtime/issues/58381, we'll be able to remove the syscall that is made for `CanSeek`, which will require adding `ESPIPE` here.

It's safer to add it, and there is no real cost to it.